src/protocols/bdx/AsyncTransferFacilitator.cpp

```diff
@@ -0,0 +1,154 @@
+                                                       chip::System::PacketBufferHandle && payload)
+{
+    VerifyOrReturnError(ec == mExchange.Get(), CHIP_ERROR_INCORRECT_STATE);
+
+    ChipLogDetail(BDX, " %s: message " ChipLogFormatMessageType " protocol " ChipLogFormatProtocolId, __FUNCTION__,
+                  payloadHeader.GetMessageType(), ChipLogValueProtocolId(payloadHeader.GetProtocolID()));
+    CHIP_ERROR err =
+        mTransfer.HandleMessageReceived(payloadHeader, std::move(payload), System::SystemClock().GetMonotonicTimestamp());
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(BDX, "failed to handle message: %" CHIP_ERROR_FORMAT, err.Format());
+    }
+
+    // Almost every BDX message will follow up with a response on the exchange. Even messages that might signify the end of a
+    // transfer could necessitate a response if they are received at the wrong time.
+    // For this reason, it is left up to the application logic to call ExchangeContext::Close() when it has determined that the
+    // transfer is finished.
+    ec->WillSendMessage();
+
+    // Get the next output event and send it to the delegate.
+    TransferSession::OutputEvent outEvent;
+    do
+    {
+        mTransfer.GetNextAction(outEvent);
+
+        if (outEvent.EventType == TransferSession::OutputEventType::kTransferTimeout)
+        {
+            OnResponseTimeout(ec);
+        }
+        else
+        {
+            HandleAsyncTransferSessionOutput(outEvent);
+        }
+    } while (outEvent.EventType != TransferSession::OutputEventType::kNone);
+
+    return err;
+}
+
+Messaging::ExchangeContext * AsyncTransferFacilitator::GetExchangeContext()
+{
+    if (mExchange)
+    {
+        return mExchange.Get();
+    }
+    return nullptr;
+}
+
+void AsyncTransferFacilitator::OnExchangeClosing(chip::Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "%s, ec: " ChipLogFormatExchange, __FUNCTION__, ChipLogValueExchange(ec));
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+}
+
+void AsyncTransferFacilitator::OnResponseTimeout(Messaging::ExchangeContext * ec)
+{
+    ChipLogDetail(BDX, "%s, ec: " ChipLogFormatExchange, __FUNCTION__, ChipLogValueExchange(ec));
+    if (mExchange)
+    {
+        mExchange.Release();
+    }
+}
+
+CHIP_ERROR AsyncResponder::PrepareForTransfer(Messaging::ExchangeContext * exchangeCtx, TransferRole role,
+                                              BitFlags<TransferControlFlags> xferControlOpts, uint16_t maxBlockSize,
+                                              System::Clock::Timeout timeout)
+{
+    VerifyOrReturnError(exchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(mTransfer.WaitForTransfer(role, xferControlOpts, maxBlockSize, timeout));
+
+    VerifyOrReturnError(!mExchange, CHIP_ERROR_INCORRECT_STATE);
+    mExchange.Grab(exchangeCtx);
+
+    return CHIP_NO_ERROR;
+}
+
+void AsyncResponder::ResetTransfer()
```

Comment:
Again, what is the point of this function?  When the transfer is done, shouldn't this object just be destroyed?

Reply:
That would make sense. i just kept existing behavior where the ota image handler's resetstate calls the rest transfer on the asyncresponder. The ota image handlers resetstate is called from a bunch of places where we error out. So this is called not only when the transfer is done.

Reply:
Basically, when we error out... shouldn't this object just get deleted?

Reply:
it should. let me see if the object is deleted when we reset state. if not ll make the changes to make sure it does.

Reply:
done.